rosdistro_reviewer/element_analyzer/rosdep.py

```diff
@@ -0,0 +1,356 @@
+                    if not isinstance(sub_rule, dict):
+                        continue
+                    if 'packages' in sub_rule:
+                        if not getattr(key, '__lines__', None):
+                            continue
+                        if key in os_installers:
+                            continue
+                        recommendation = Recommendation.DISAPPROVE
+                        problems.add(
+                            'One or more explicitly provided installer is not '
+                            'supported by rosdep')
+                        annotations.append(Annotation(
+                            file,
+                            range(
+                                key.__lines__.start,
+                                sub_rule.__lines__.stop),
+                            f"Installer '{key}' is not supported for '{os}'"))
+                        continue
+                    for installer, sub_sub_rule in sub_rule.items():
+                        if not getattr(installer, '__lines__', None):
+                            continue
+                        if installer in os_installers:
+                            continue
+                        recommendation = Recommendation.DISAPPROVE
+                        problems.add(
+                            'One or more explicitly provided installer is not '
+                            'supported by rosdep')
+                        annotations.append(Annotation(
+                            file,
+                            range(
+                                installer.__lines__.start,
+                                sub_sub_rule.__lines__.stop),
+                            f"Installer '{installer}' is not supported "
+                            f"for '{os}'"))
+
+    if problems:
+        message = '\n- '.join([
+            'There are problems with explicitly provided installers:',
+        ] + sorted(problems))
+    else:
+        message = 'Installers for new rosdep rules are valid'
+
+    criteria.append(Criterion(recommendation, message))
+
+
+def _get_changed_rosdeps(
+    path: Path,
+    target_ref: Optional[str] = None,
+    head_ref: Optional[str] = None,
+) -> Tuple[Optional[Dict[str, int]], Optional[Dict[str, Any]]]:
+    rosdep_files = [
+        str(p.relative_to(path))
+        for p in path.glob('rosdep/*.yaml')
+    ]
+    if not rosdep_files:
+        logger.info('No rosdep files were found in the repository')
+        return None, None
+
+    changes = get_changed_yaml(
+        path, rosdep_files, target_ref=target_ref, head_ref=head_ref)
+    if not changes:
+        logger.info('No rosdep files were modified')
+        return None, None
+
+    rosdep_changes = {}
+    key_counts: Dict[str, int] = {}
+    for rosdep_file, data in changes.items():
+        if data:
+            for key in data.keys():
+                key_counts.setdefault(key, 0)
+                key_counts[key] += 1
```

Comment:
These key counts only consider files touched in the changes but the checks above seem to assume that key counts are accurate for the entire rosdep db. My interpretation of the key count check is that it makes sure that keys aren't added that already existed OR were not added twice. But I think because of this it only does the latter. Am I missing something?

Reply:
This is a good thing to call out - it's super unclear and was more of an afterthought on the implementation.

If `get_changed_yaml` detects no changes to any of the files it is scoped to, it will return `None`. If ANY of the files were changed, it will return all of the loaded content from all of the files with isolated annotations, meaning that any "changed" data in any of the files will have non-None `__lines__`. So we should have all of the files under `rosdep/*.yaml` at this point.

After we count the keys, the irrelevant data is culled by the call to `prune_changed_yaml` so that the later checks don't need to process so much data. The checks can *handle* the irrelevant data and won't consider it during checks, it's just an optimization to drop it after we're done with it so all the checks don't need to iterate over it.